File: other/cracking-the-coding-interview/ch1/1.1_is_unique/is_unique.go
package main

import "fmt"

func isUnique(input string) bool {
	seen := make(map[rune]struct{})
	for _, r := range input {
		if _, ok := seen[r]; ok {
			return false
		} else {
			seen[r] = struct{}{}
		}
	}

	return true
}

func main() {
	inputs := []string{"abcde", "hello", "apple", "kite", "padle"}
	outputs := []bool{true, false, false, true, true}
	var sol bool
	var ans bool

	for i, inp := range inputs {
		sol = isUnique(inp)
		ans = outputs[i]

		if sol != ans {
			fmt.Println(inp, sol, ans)
		}
	}
}

File: other/cracking-the-coding-interview/ch1/1.1_is_unique/is_unique.py

class Solution:
    def is_unique(self, s):
        seen = {}
        for i in s:
            if i in seen:
                return False
            else:
                seen[i] = True

        return True


if __name__ == '__main__':
    inputs = ["abcde", "hello", "apple", "kite", "padle"]
    outputs = [True, False, False, True, True]
    sol = Solution()

    for i in range(len(inputs)):
        if sol.is_unique(inputs[i]) != outputs[i]:
            print(inputs[i], outputs[i], sol.is_unique(inputs[i]))
